plugins/ccgx/fu-ccgx-hpi.c

```diff
@@ -0,0 +1,1633 @@
+		return FALSE;
+	}
+
+	device_data->fw_meta_valid = TRUE;
+	g_usleep (HPI_CMD_ENTER_FLASH_MODE_DELAY_US);
+	return TRUE;
+}
+
+/**
+ * fu_ccgx_hpi_cmd_enter_flash_mode:
+ * @device: #FuDevice
+ * @hpi_handle HPI handle
+ * @error: a #GError or %NULL
+ *
+ * Enter flash mode in PD I2C Device
+ *
+ * Returns: %TRUE for success
+*/
+gboolean
+fu_ccgx_hpi_cmd_enter_flash_mode (FuDevice *device, CyHPIHandle *hpi_handle,  GError **error)
+{
+	CyPDResp hpi_event = 0;
+
+	g_return_val_if_fail (hpi_handle != NULL, FALSE);
+
+	(void)fu_ccgx_hpi_clear_all_event(device,hpi_handle,HPI_CMD_COMMAND_CLEAR_EVENT_TIME_MS,error);
+
+	if (!fu_ccgx_hpi_enter_flash_mode (device, hpi_handle, error)) {
+		g_warning("enter flash mode cmd	error");
+		return FALSE;
+	}
+
+	if (!fu_ccgx_hpi_get_event (device, hpi_handle, HPI_REG_SECTION_DEV, &hpi_event, HPI_CMD_COMMAND_RESPONSE_TIME_MS ,error)) {
+		g_warning("enter flash mode resp error");
+		return FALSE;
+	}
+
+	if (hpi_event != CY_PD_RESP_SUCCESS) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "enter flash mode resp code error:	0x%x", hpi_event);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/**
+ * fu_ccgx_hpi_cmd_leave_flash_mode:
+ * @device: #FuDevice
+ * @hpi_handle HPI handle
+ * @error: a #GError or %NULL
+ *
+ * Leave flash mode in PD I2C Device
+ *
+ * Returns: %TRUE for success
+*/
+gboolean
+fu_ccgx_hpi_cmd_leave_flash_mode (FuDevice *device, CyHPIHandle *hpi_handle, GError	**error)
+ {
+	CyPDResp hpi_event = 0;
+
+	g_return_val_if_fail (hpi_handle != NULL, FALSE);
+	(void)fu_ccgx_hpi_clear_all_event(device,hpi_handle,HPI_CMD_COMMAND_CLEAR_EVENT_TIME_MS,error);
+
+	if (!fu_ccgx_hpi_leave_flash_mode (device, hpi_handle, error)) {
+		 g_warning("leave flash mode cmd error");
+		 return	FALSE;
+	}
+
+	if (!fu_ccgx_hpi_get_event (device, hpi_handle, HPI_REG_SECTION_DEV, &hpi_event, HPI_CMD_COMMAND_RESPONSE_TIME_MS ,error)) {
+		g_warning("leave flash mode resp error");
+		return FALSE;
+	}
+
+	if (hpi_event != CY_PD_RESP_SUCCESS) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "leave flash mode resp code error: 0x%x", hpi_event);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/**
+ * fu_ccgx_hpi_cmd_write_flash:
+ * @device: #FuDevice
+ * @hpi_handle HPI handle
+ * @row_num row number
+ * @data flash row data
+ * @size size of row data
+ * @error: a #GError or %NULL
+ *
+ * Write Flash data to PD I2C Device
+ *
+ * Returns: %TRUE for success
+*/
+gboolean
+fu_ccgx_hpi_cmd_write_flash (FuDevice *device, CyHPIHandle *hpi_handle, guint16 row_num, guint8 *data, guint16 size, GError **error)
+{
+	CyPDResp hpi_event = 0;
+	guint16 reg_addr = 0;
+
+	g_return_val_if_fail (hpi_handle != NULL, FALSE);
+	g_return_val_if_fail (data != NULL, FALSE);
+	g_return_val_if_fail (size > 0, FALSE);
+
+	(void)fu_ccgx_hpi_clear_all_event(device,hpi_handle,HPI_CMD_COMMAND_CLEAR_EVENT_TIME_MS,error);
```

Comment:
should return false and use g_prefix_error?

Reply:
OK. I will make fu_ccgx_hpi_clear_all_events get process GError and get return value when it has error.